opa-iptables/pkg/iptables/rule.go

```diff
@@ -0,0 +1,219 @@
+	// A protocol name from /etc/protocols is also allowed.
+	// A ! argument before the protocol inverts the test.
+	// The number zero is equivalent to all.
+	// all will match with all protocols and is taken as default when this option is omitted.
+	Protocol string `json:"protocol"`
+
+	// TCP flags specification.
+	// tcp_flags expects a dict with the two keys flags and flags_set.
+	TCPFlags tcpFlags `json:"tcp_flags"`
+
+	// ctstate is a list of the connection states to match in the conntrack module.
+	// Possible states are INVALID, NEW, ESTABLISHED, RELATED, UNTRACKED, SNAT, DNAT
+	Ctstate []string `json:"ctstate"`
+
+	// Specifies a match to use, that is, an extension module that tests for a specific property.
+	// The set of matches make up the condition under which a target is invoked.
+	// Matches are evaluated first to last if specified as an array and work in short-circuit fashion,
+	// i.e. if one extension yields false, evaluation will stop.
+	Match []string `json:"match"`
+}
+
+type tcpFlags struct {
+	// List of flags you want to examine.
+	Flags []string `json:"flags"`
+	// Flags to be set.
+	FlagsSet []string `json:"flags_set"`
+}
+
+func addParam(buf *bytes.Buffer,param,flag string) {
+	if param != "" {
+		if param[0] == '!' {
+			fmt.Fprintf(buf," ! %s %s",flag,param[1:])
+		}else{
+			fmt.Fprintf(buf," %s %s",flag,param)
+		}
+	}
+}
+
+func addParams(buf *bytes.Buffer,params []string, flag string) {
+	for _,param := range params {
+		addParam(buf,param,flag)
+	}
+}
+
+func addMatch(buf *bytes.Buffer,param string) {
+	if param != "" {
+		fmt.Fprintf(buf," -m %s",param)
+	}
+}
+
+func addComment(buf *bytes.Buffer,comment string) {
+	if comment != "" {
+		addMatch(buf,"comment")
+		fmt.Fprintf(buf," --comment \"%s\"",comment)
+	}
+}
+
+func addIPRange(buf *bytes.Buffer,matchs []string, sourceRange, destinationRange string) {
+	for _,match := range matchs {
+		if match == "iprange" {
+			addParam(buf,sourceRange,"--src-range")
+			addParam(buf,destinationRange,"--dst-range")
+			return
+		}		
+	}
+	if sourceRange != "" || destinationRange != "" {
+		addMatch(buf,"iprange")
+		addParam(buf,sourceRange,"--src-range")
+		addParam(buf,destinationRange,"--dst-range")
+	}
+}
+
+func addTCPFlags(buf *bytes.Buffer,tf tcpFlags) {
+	if len(tf.Flags) > 0 && len(tf.FlagsSet) > 0 {
+		addParam(buf,strings.Join(tf.Flags,",")+" "+strings.Join(tf.FlagsSet,","),"--tcp-flags")
+	}
+}
+
+func addCTState(buf *bytes.Buffer,matchs,states []string) {
+	for _,match := range matchs {
+		if match == "conntrack" {
+			addParam(buf,strings.Join(states,","),"--ctstate")
+			return
+		}else if match == "state" {
+			addParam(buf,strings.Join(states,","),"--ctstate")
+			return
+		}
+	}
+	if len(states) > 0 {
+		addMatch(buf,"conntrack")
+		addParam(buf,strings.Join(states,","),"--ctstate")
+	}
+}
+
+// Construct IPTable rule from struct
+func (r *Rule) Construct() string {
+	var rule bytes.Buffer
```

Comment:
I would suggest calling this local variable something else since the struct is also called Rule (with capital R). Otherwise we have rule and Rule in the same function